File: fft/go/fft_perf.go
package main

import (
    "fmt"
    "math"
    "time"
)

const N = 1 << 20

type Complex struct {
    re, im float64
}

func add(a, b Complex) Complex {
    return Complex{a.re + b.re, a.im + b.im}
}

func sub(a, b Complex) Complex {
    return Complex{a.re - b.re, a.im - b.im}
}

func mul(a, b Complex) Complex {
    return Complex{
        a.re*b.re - a.im*b.im,
        a.re*b.im + a.im*b.re,
    }
}

func fft(x []Complex) {
    n := len(x)
    if n <= 1 {
        return
    }

    even := make([]Complex, n/2)
    odd := make([]Complex, n/2)
    for i := 0; i < n/2; i++ {
        even[i] = x[2*i]
        odd[i] = x[2*i+1]
    }

    fft(even)
    fft(odd)

    for k := 0; k < n/2; k++ {
        angle := -2 * math.Pi * float64(k) / float64(n)
        twiddle := Complex{math.Cos(angle), math.Sin(angle)}
        t := mul(twiddle, odd[k])
        x[k] = add(even[k], t)
        x[k+n/2] = sub(even[k], t)
    }
}

func main() {
    fmt.Printf("Generating %d-point complex FFT input...\n", N)
    fmt.Println("Technology: Go")

    data := make([]Complex, N)
    for i := 0; i < N; i++ {
        data[i] = Complex{float64(i % 256), 0.0}
    }

    start := time.Now()
    fft(data)
    elapsed := time.Since(start).Seconds()

    fmt.Printf("FFT %d points: %.6f seconds\n", N, elapsed)
}

File: fft/java/FFTPerf.java
public class FFTPerf {

    static class Complex {
        double re, im;

        Complex(double re, double im) {
            this.re = re;
            this.im = im;
        }

        Complex add(Complex other) {
            return new Complex(this.re + other.re, this.im + other.im);
        }

        Complex sub(Complex other) {
            return new Complex(this.re - other.re, this.im - other.im);
        }

        Complex mul(Complex other) {
            return new Complex(
                this.re * other.re - this.im * other.im,
                this.re * other.im + this.im * other.re
            );
        }
    }

    static void fft(Complex[] x) {
        int n = x.length;
        if (n <= 1) return;

        Complex[] even = new Complex[n / 2];
        Complex[] odd = new Complex[n / 2];
        for (int i = 0; i < n / 2; i++) {
            even[i] = x[2 * i];
            odd[i] = x[2 * i + 1];
        }

        fft(even);
        fft(odd);

        for (int k = 0; k < n / 2; k++) {
            double angle = -2 * Math.PI * k / n;
            Complex twiddle = new Complex(Math.cos(angle), Math.sin(angle)).mul(odd[k]);
            x[k] = even[k].add(twiddle);
            x[k + n / 2] = even[k].sub(twiddle);
        }
    }

    public static void main(String[] args) {
        final int N = 1 << 20;
        System.out.println("Generating " + N + "-point complex FFT input...");
        System.out.println("Technology: Java");

        Complex[] data = new Complex[N];
        for (int i = 0; i < N; i++) {
            data[i] = new Complex(i % 256, 0.0);
        }

        long start = System.nanoTime();
        fft(data);
        long end = System.nanoTime();

        double elapsed = (end - start) / 1e9;
        System.out.printf("FFT %d points: %.6f seconds%n", N, elapsed);
    }
}

